README.md
# UserIQ iOS SDK Integration Document

[![CocoaPods Compatible](https://img.shields.io/cocoapods/v/UserIQ.svg)](https://img.shields.io/cocoapods/v/UserIQ.svg)
[![Carthage Compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)](https://github.com/Carthage/Carthage)
[![Platform](https://img.shields.io/cocoapods/p/UserIQ.svg?style=flat)](http://cocoapods.org/pods/UserIQ)

This guide will provide you with step by step details on how to integrate the SDK in just a few minutes. The following steps outline the integration process in details.

## Steps to integrate the sdk to your Xcode - iOS project

- [Step 1 : Download the UserIQ framework](https://github.com/useriq-com/ios-sdk#step-1--download-the-useriq-framework)
- [Step 2 : Attach UserIQ framework to your iOS project](https://github.com/useriq-com/ios-sdk#step-2--attach-useriq-framework-to-your-ios-project)
- [Step 3 : Initiate the UserIQ framework](https://github.com/useriq-com/ios-sdk#step-3--initiate-the-useriq-framework)
- [Step 4 : Set the user after login](https://github.com/useriq-com/ios-sdk#step-4--set-the-user-after-login)
- [Step 5 : Add the custom parameters (optional)](https://github.com/useriq-com/ios-sdk#step-5--add-the-custom-parameters-optional)
- [Step 6 : Logout](https://github.com/useriq-com/ios-sdk#step-6--logout)

If you are using Cocoapods or Carthage, skip Step 1 & Step 2, go directly to [Step 3](#Step-3--Initiate-the-useriq-framework).

### Cocoapods

Add

```
  pod 'UserIQ', '~> 2.9'
```

to your Podfile.
In terminal, go to the folder of your project and enter `pod install`. Open <your-project-name.xcworkspace>

### Carthage

Add

```
  github "useriq-com/ios-sdk"
```

to your Cartfile. In terminal, go to the folder of your project and enter `carthage update`

### Step 1 : Download the UserIQ framework

Download the UserIQ sdk from [here](https://github.com/useriq-com/ios-sdk/releases)

Alternatively you can clone the UserIQ github repo

```
  git clone https://github.com/useriq-com/ios-sdk.git
```

The **_UserIQ.framework_** is to be copied into your project folder.

### Step 2 : Attach UserIQ framework to your iOS project

- Click on the **_project_** folder on the project navigator panel.
- Click on the application under **_Targets_** and go to the **_General_** tab as shown in image below
- Click on **_+_** button under **_Embedded Binaries_** as shown in image below.
- Click on **_Add Other_** from the dialog that appears.
- Select **_UserIQ.framework_** from the project folder.
- If prompted with a dialog that asks whether to copy the framework, click on **_Yes_**.

### Step 3 : Initiate the UserIQ framework

- In your AppDelegate file, import the UserIQ framework

  Swift :

  ```Swift
    import UserIQ
  ```

  Objective-C:

  ```objc
    #import <UserIQ/UserIQ.h>
  ```

- In the `didFinishLaunchingWithOptions:` method initiate the UserIQ sdk using the initMethod as shown below

  Swift :

  ```Swift
    func application(_application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
      UserIQSDK.sharedInstance().initWithAPIKey("<YOUR-API-KEY>")
    }
  ```

  Objective-C :

  ```objc
    -(BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
      // Override point for customization after application launch.
      [[UserIQSDK sharedInstance] initWithAPIKey:@"<YOUR-API-KEY>"];
    }
  ```

  Note - The **_API-KEY_** can be obtained from the integration page of your app in the UserIQ dashboard.

  This API will initialise the SDK with an anonymous user.

### Step 4 : Set the user after login

- Once user has logged in, the anonymous user can be changed to the current user using the `setUser` API as shown below. Import the UserIQ SDK in the file (similar to in the previous step) where you receive the user info.

  Swift :

  ```Swift
    UserIQSDK.sharedInstance().setUserId("EMP124",
                                      name: "Alex",
                                      email: "alex@useriq.com",
                                      accountId: "1",
                                      accountName: "Acme Corp",
                                      signupDate: "2017-11-29",
                                      andParameters: ["location":"Atlanta"])
  ```

  Objective-C:

  ```objc
    [[UserIQSDK sharedInstance] setUserId:@"EMP124"
                                     name:@"Alex"
                                    email:@"alex@useriq.com"
                                accountId:@"1"
                              accountName:@"Acme Corp"
                               signupDate:@"2017-11-29"
                            andParameters:@{@"location":@"Atlanta"}];
  ```

  Note - Make sure this api is called on every launch of the app if the app keep users logged in until user manually logs out.

### Step 5 : Add the custom parameters (optional)

- Add any custom attributes about the user in andParameters(String key, String value)

  Swift :

  ```Swift
     andParameters: ["location":"Atlanta", "Foo":"Bar", "Foo1": "Bar1"]
  ```

  Objective-C :

  ```objc
     andParameters:@{@"location":@"Atlanta", @"Foo":@"Bar", @"Foo1":@"Bar1"}
  ```

### Step 6 : Logout

- If a user logs out, the user can be reset to anonymous user just by calling the `logout` API. Make sure this method is called when the user logs out, so that login screen tracking and other information not related to the user does not get linked to the user.

  Swift :

  ```Swift
    UserIQSDK.sharedInstance().logout()
  ```

  Objective-C :

  ```objc
    [[UserIQSDK sharedInstance] logout];
  ```

# API & USAGE

For more details on API & their usage, please refer to [wiki page](https://github.com/useriq-com/ios-sdk/wiki)


CHANGELOG.md
# UserIQ iOS SDK Changelog

All notable changes in iOS SDK are documented here. We adhere to
[Semantic Versioning](http://semver.org/spec/v2.0.0.html).

### [2.9.0] - 2021-06-22

#### Fixed

- Fixed react-native screen freezing on alert window (only Appcontent window will be considered as valid Window for tagging & tours)

---

### [2.8.0] - 2020-09-01

#### Added

- Multiple lines support for Tooltip Description
- tzOffset support for MobileAppEvents

---

### [2.7.1] - 2020-07-15

#### Fixed

- Proper cropping of images in campaigns
- Headerview fix for campaign with webview

---

### [2.7.0] - 2020-06-29

#### Added 

- Launching InApp Modal from another InApp Modal

#### Fixed

- Improved Screen Identification for React-Native 
- iOS 13 fix for UIWindow  

---

### [2.6.3] - 2020-02-28

#### Added

- Additional Logging

---

### [2.6.2] - 2019-12-20

#### Fixed

- Full screen support for `Help Center` in iOS 13
- `Help Center` welcome text alignment set to left

---

### [2.6.1] - 2019-12-12

#### Fixed

- Optimised click tracking (Eliminates traversal)
- Ignored touch tracking for `UIScrollView` and `UITextField`

---

### [2.6.0] - 2019-12-04

#### Added

- Logout API.
- SDK initialised with anonymous user when initWithAPIKey is called without setUser
- Added provision for SDK kill switch.

#### Updated

- Store additional parameters for same user in case of multiple `setUser` calls

#### Removed

- deactivateSDK API
- isOffline API

---

### [2.5.2] - 2019-11-28

#### Updated

- Improved debug logging

---

### [2.5.1] - 2019-11-07

#### Updated

- Reduced in-app message delay from 3000 ms to 250 ms

---

### [2.5.0] - 2019-10-09

#### Updated

- Increased timeout for network calls
- Improved performance in live previews
- Minor bug fixes & performance improvements

---

### [2.4.2] - 2019-08-30

#### Added

- Animation support for inapp messages

#### Updated

- App logo capture in Xamarin apps

---

### [2.4.1] - 2019-08-08

#### Updated

- Exposed `APP_FOREGROUND` event to dashboard

---

### [2.4.0] - 2019-07-25

#### Added

- Comment box support in NPS & Star campaigns
- New campaign button action `OPEN_WEBVIEW`. Opens a url inside webview in the app.
- Load balancer support for SDK endpoint

#### Updated

- Improved retry connection mechanism

---

### [2.3.2] - 2019-07-18

#### Fixed

- Minor bug fixes for Xamarin support

---

### [2.3.1] - 2019-07-11

#### Fixed

- Back button fix in help center in iPhones with notch

---

### [2.3.0] - 2019-07-09

#### Added

- New campaign button action support: `Open Question` & `Start Walkthrough`

#### Updated

- Improved UI for helpcenter

---

### [2.2.0] - 2019-06-17

#### Added

- Automatic/Manual tooltip positioning support (2 positions)
- Automatic/Manual beacon and number tooltip positioning support (9 positions)
- Support for `x` and `y` offsets for tooltips
- Ability to make walkthorugh wait for the next step for a specific amount of time
- Keyboard support for simulators (on Dashboard)
- Copy/Paste support for simulators (on Dashboard)

#### Updated

- Improved screen identification (5x to 8x faster)

---

### [2.1.0] - 2019-05-30

#### Added

- Landscape Orientation support (on Dashboard)
- Performance tracking for screen change

#### Updated

- Improved screen identification (4x to 10x faster)
- Added event tracking for campaings (walkthrough & campaigns)

#### Fixed

- Fixed React Native screen identification
- Fixed React Native click tracking
- Other minor bug fixes

---

### [2.0.5] - 2019-05-11

#### Updated

- Reuse existing websocket when setUser is called

---

### [2.0.4] - 2019-04-18

#### Added

- Bitcode enabled framework

---

### [2.0.3] - 2019-04-16

#### Added

- New api for setting user. This can be used to log in and logout user. Keep a standard user info for logged out users for tracking purposes.

---

### [2.0.2] - 2019-04-11

#### Fixed

- Bug Fix : FAB button pushed to front even if new view is added in top layer

---

### [2.0.1] - 2019-03-27

#### Fixed

- Minor bug fixes

---

### [2.0.0] - 2018-12-18

#### Added

- NPS Surveys (3 types)
- Rating Surveys (3 types)
- Walkthroughs (Beacons, Numbered & Tooltips)
  - Both single step and multiple steps in the same screen.
  - Cross screen walkthroughs with step triggers
- Helpcenter
  - Supports rich text format
  - Ability to add images (auto cached on device)
  - Ability to export walkthrough’s into questions
- Contextual Help
  - Configurable guide on per screen basis
  - Allow manual triggers of FAQ's & Walkthroughs

---

### [1.0.3] - 2018-07-20

#### Added

- Host support

---

### [1.0.2] - 2018-07-20

#### Added

- Box shadow & Text shadow support
- Modal related events

---

### [1.0.1] - 2018-07-02

#### Fixed

- Bug Fix : Text color change for in-app modal heading and body

---

### [1.0.0] - 2018-06-29

#### Added

- Initial Release of UserIQ iOS SDK

---


Sample App/UserIQ.framework/Headers/UserIQSDK.h
//
//  UserIQSDK.h
//  UserIQ
//
//  Created by Aravind G S on 26/05/17.
//  Copyright © 2017 UserIQ. All rights reserved.
//

#import <Foundation/Foundation.h>
#import <UIKit/UIKit.h>


@interface UserIQSDK : NSObject


/**
 Create a global instance

 @return global instace
 */
+ (UserIQSDK * _Nonnull) sharedInstance;

/**
 Only required for UserIQ developers. To be called before SDK initialization

 @param host The host to be connected to.
 */
- (void) setHost:(NSString * _Nonnull)host;

/**
 Method to set api key. Initiates the SDK with an anonymous user

 @param apiKey api key available from useriq dashboard. Will setup an anonymous user and initiate the SDK. The actual user info can be passed after the user logins using the  -setUserId: name: email: accountId: accountName: signupDate: andParameters: API
 */
- (void) initWithAPIKey:(NSString * _Nonnull)apiKey;


/**
 Method to initialize SDK with all user info. If user info is not available initially, empty values has to be passed. And later when the user info is available the setUser method can be called

 @param token api key available from useriq dashboard
 @param userId userid of the user
 @param name name of the user
 @param email email of the user
 @param accId account of the user
 @param accName account name of the user
 @param signupDate signup date of the user
 @param parameters additional parameters as key-value pairs relevant to user
 */
- (void) initWithAPIKey:(NSString * _Nonnull)token
                 userId:(NSString * _Nonnull)userId
                   name:(NSString * _Nonnull)name
                  email:(NSString * _Nonnull)email
              accountId:(NSString * _Nonnull)accId
            accountName:(NSString * _Nonnull)accName
             signupDate:(NSString * _Nullable)signupDate
          andParameters:(NSDictionary <NSString *, NSString*> * _Nullable)parameters  __attribute__((deprecated));

/**
 Method to set/reset user.
 
 If you are setting the user for the first time after the app launch using this method, it should be only called after api key has been already set using one of the `initWithAPIKey` methods.

 @param userId userid of the user
 @param name name of the user
 @param email email of the user
 @param accId account of the user
 @param accName account name of the user
 @param signupDate of the user
 @param parameters additional parameters as key-value pairs relevant to user
 */

- (void) setUserId:(NSString * _Nonnull)userId
              name:(NSString * _Nonnull)name
             email:(NSString * _Nonnull)email
         accountId:(NSString * _Nonnull)accId
       accountName:(NSString * _Nonnull)accName
        signupDate:(NSString * _Nullable)signupDate
     andParameters:(NSDictionary <NSString *, NSString*> * _Nullable)parameters;


/**
Method to be called when user logs out. The user is reset to anonymous user automatically here.
 */
- (void) logout;

/**
 Get the version of the UserIQ SDK

 @return version
 */
+ (NSString * _Nonnull) version;

/**
 Method to disable FAB button
 */
- (void)disableFAB;

/**
 Presents the contextual help for the current screen.

 @return YES if ctxHelp is presented. NO if no ctxHelp is available.
 */
- (BOOL) showCtxHelp;


/**
 Presents the help center
 */
- (void) showHelpCentre;

@end


Sample App/Sample App/AppDelegate.m
//
//  AppDelegate.m
//  Sample App
//
//  Created by Aravind G S on 29/06/18.
//  Copyright © 2018 useriq. All rights reserved.
//

#import "AppDelegate.h"

@interface AppDelegate ()

@end

@implementation AppDelegate


- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
    // Override point for customization after application launch.
    [[UserIQSDK sharedInstance] initWithAPIKey:@"9d36faecef6a1efa6d808509785d4fc53a2bf37f"];
    [[UserIQSDK sharedInstance] setUserId:@"EMP124"
                                     name:@"Alex"
                                    email:@"alex@useriq.com"
                                accountId:@"1"
                              accountName:@"Acme Corp"
                               signupDate:@"2017-04-21"
                            andParameters:@{@"location":@"Atlanta"}];
    return YES;
}


- (void)applicationWillResignActive:(UIApplication *)application {
    // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
    // Use this method to pause ongoing tasks, disable timers, and invalidate graphics rendering callbacks. Games should use this method to pause the game.
}


- (void)applicationDidEnterBackground:(UIApplication *)application {
    // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
    // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
}


- (void)applicationWillEnterForeground:(UIApplication *)application {
    // Called as part of the transition from the background to the active state; here you can undo many of the changes made on entering the background.
}


- (void)applicationDidBecomeActive:(UIApplication *)application {
    // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
}


- (void)applicationWillTerminate:(UIApplication *)application {
    // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
}


@end
